AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/MutationQueue/ProcessMutationErrorFromCloudOperationTests.swift

```diff
@@ -46,19 +52,719 @@ class ProcessMutationErrorFromCloudOperationTests: XCTestCase {
+        let mutationEvent = try MutationEvent(model: localPost, mutationType: .update)
+        guard let graphQLResponseError = try getGraphQLResponseError(withRemote: remotePost,
+                                                                     deleted: true,
+                                                                     version: 2) else {
+            XCTFail("Couldn't get GraphQL response with remote post")
+            return
+        }
+        let expectCompletion = expectation(description: "Expect to complete error processing")
+        let completion: (Result<MutationEvent?, Error>) -> Void = { result in
+            guard case .success(let mutationEventOptional) = result,
+                let mutationEvent = mutationEventOptional else {
+                XCTFail("Should have been successful")
+                return
+            }
+            XCTAssertEqual(mutationEvent.mutationType, "delete")
+            XCTAssertEqual(mutationEvent.modelId, localPost.id)
+            expectCompletion.fulfill()
+        }
+
+        let modelDeletedEvent = expectation(description: "model deleted event")
+        let metadataSavedEvent = expectation(description: "metadata saved event")
+        let storageAdapter = MockSQLiteStorageEngineAdapter()
+        storageAdapter.shouldReturnErrorOnDeleteMutation = false
+        storageAdapter.responders[.deleteUntypedModel] = DeleteUntypedModelCompletionResponder { _ in
+            modelDeletedEvent.fulfill()
+        }
+        storageAdapter.responders[.saveModelCompletion] =
+            SaveModelCompletionResponder<MutationSyncMetadata> { metadata, completion in
+            XCTAssertEqual(metadata.deleted, true)
+            XCTAssertEqual(metadata.version, 2)
+            metadataSavedEvent.fulfill()
+            completion(.success(metadata))
+        }
+
+        let expectHubEvent = expectation(description: "Hub is notified")
+        let hubListener = Amplify.Hub.listen(to: .dataStore) { payload in
+            if payload.eventName == "DataStore.syncReceived" {
+                expectHubEvent.fulfill()
+            }
+        }
+        let operation = ProcessMutationErrorFromCloudOperation(dataStoreConfiguration: .default,
+                                                               mutationEvent: mutationEvent,
+                                                               api: mockAPIPlugin,
+                                                               storageAdapter: storageAdapter,
+                                                               error: graphQLResponseError,
+                                                               completion: completion)
+
+        let queue = OperationQueue()
+        queue.addOperation(operation)
+
+        wait(for: [modelDeletedEvent, metadataSavedEvent, expectHubEvent, expectCompletion],
+             timeout: defaultAsyncWaitTimeout)
```

Comment:
It might be more readable if we be strict about the order in which expectations finish,  we can break them up into separate lines and have multiple wait statements.

Reply:
really like this idea, makes it clear which fulfill is supposed to be executed first